include/zephyr/drivers/gpio.h

```diff
@@ -417,6 +417,247 @@ struct gpio_dt_spec {
+ *	};
+ * @endcode
+ *
+ * @param node_id GPIO controller node identifier.
+ * @param ngpios number of GPIOs.
+ * @return the bitmask of reserved gpios
+ */
+#define GPIO_DT_RESERVED_RANGES_NGPIOS(node_id, ngpios)			       \
+	((gpio_port_pins_t)						       \
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, gpio_reserved_ranges),	       \
+		(GENMASK64(BITS_PER_LONG_LONG - 1, ngpios)		       \
+		| FOR_EACH_FIXED_ARG(Z_GPIO_GEN_RESERVED_RANGES_COND,	       \
+			(|),						       \
+			node_id,					       \
+			LIST_DROP_EMPTY(Z_SPARSE_LIST_ODD_NUMBERS))),	       \
+		(0)))
+
+/**
+ * @brief Makes a bitmask of reserved GPIOs from the @p "gpio-reserved-ranges"
+ *        and @p "ngpios" DT properties values
+ *
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of reserved gpios
+ */
+#define GPIO_DT_RESERVED_RANGES(node_id)				       \
+	GPIO_DT_RESERVED_RANGES_NGPIOS(node_id, DT_PROP(node_id, ngpios))
+
+/**
+ * @brief Makes a bitmask of reserved GPIOs from a DT_DRV_COMPAT instance's
+ *        @p "gpio-reserved-ranges" property and @p "ngpios" argument
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @return the bitmask of reserved gpios
+ * @param ngpios  number of GPIOs
+ * @see GPIO_DT_RESERVED_RANGES()
+ */
+#define GPIO_DT_INST_RESERVED_RANGES_NGPIOS(inst, ngpios)		       \
+		GPIO_DT_RESERVED_RANGES_NGPIOS(DT_DRV_INST(inst), ngpios)
+
+/**
+ * @brief Make a bitmask of reserved GPIOs from a DT_DRV_COMPAT instance's GPIO
+ *        @p "gpio-reserved-ranges" and @p "ngpios" properties
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @return the bitmask of reserved gpios
+ * @see GPIO_DT_RESERVED_RANGES()
+ */
+#define GPIO_DT_INST_RESERVED_RANGES(inst)				       \
+		GPIO_DT_RESERVED_RANGES(DT_DRV_INST(inst))
+
+/**
+ * @brief Makes a bitmask of allowed GPIOs from DT @p "gpio-reserved-ranges"
+ *        property and @p "ngpios" argument
+ *
+ * This macro is paired with GPIO_DT_RESERVED_RANGES_NGPIOS(), however unlike
+ * the latter, it returns a bitmask of ALLOWED gpios.
+ *
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+ *	a {
+ *		compatible = "some,gpio-controller";
+ *		ngpios = <32>;
+ *		gpio-reserved-ranges = <0 8>, <9 5>, <15 16>;
+ *	};
+ *
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *	struct some_config {
+ *		uint32_t port_pin_mask;
+ *	};
+ *
+ *	static const struct some_config dev_cfg = {
+ *		.port_pin_mask = GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(
+ *					DT_LABEL(a), 32),
+ *	};
+ * @endcode
+ *
+ * This expands to:
+ *
+ * @code{.c}
+ *	struct some_config {
+ *		uint32_t port_pin_mask;
+ *	};
+ *
+ *	static const struct some_config dev_cfg = {
+ *		.port_pin_mask = 0x80004100,
+ *				// 0b1000 0000 0000 0000 0100 0001 00000 000
+ *	};
+ * @endcode
+ *
+ * @param node_id GPIO controller node identifier.
+ * @param ngpios  number of GPIOs
+ * @return the bitmask of allowed gpios
+ */
+#define GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(node_id, ngpios)		       \
```

Comment:
Should the two `_EXC` macros be marked as internal? Is there a use case for these outside of calling `GPIO_PORT_PIN_MASK_FROM_DT_NODE()`?

Reply:
If you will check my reverted modification of gpio drivers, this macro can be useful for the 'nonstandart' gpio drivers, like your gpio_xlnx_axi.c